### ISSUE TITLE
Support for Windows purl type

### ISSUE THREAD

**Issue by rhdesmond** [package-url/purl-spec#255]
I noticed that most common operating systems and language types are supported, but not Windows.  Would it be possible to add Windows as a supported Purl type, perhaps with the signifier `win`

**Comment by stevespringett** [comment]
I think "windows" is too generic, however, support for the Microsoft Store would be a welcome addition. There may be other types as well. Sorry, not a Windows person.

Did you want to propose the purl type definition for the Microsoft Store?

**Comment by alowayed** [comment]
Not the store, but for Windows exe packages.

For example, how would one represent the MySQL community server package for Windows found at: https://dev.mysql.com/downloads/mysql/ . `msi` might be a better type than `windows` since the package is installed / managed by the windows installer:

![Screenshot 2023-10-25 at 3 07 34 PM](https://github.com/package-url/purl-spec/assets/10391939/ff3c9049-1b80-4f1f-af12-9482a48cf436)

Contrast this to the RHEL version of MySQL that is managed through RPM and does have a specific PURL (e.g. `pkg:rpm/ol/mysql-community-server-minimal@8.2`):
![Screenshot 2023-10-25 at 3 10 08 PM](https://github.com/package-url/purl-spec/assets/10391939/a1a10d2e-269e-494d-8190-1c9887230d08)

Also not a Windows person so correct me if I'm wrong.

**Comment by prabhu** [comment]
Perhaps use swid with purl instead of reinventing another id?

**Comment by bureado** [comment]
> For example, how would one represent the MySQL community server package for Windows found at: https://dev.mysql.com/downloads/mysql/ . `msi` might be a better type than `windows` since the package is installed / managed by the windows installer:

I think `generic` would be appropriate in this case.

For OS, see: https://github.com/package-url/purl-spec/pull/161

For WinGet, Chocolatey, Store, etc., I'm not an expert either but I do think those could be distinct types.

**Comment by prabhu** [comment]
https://dev.mysql.com/get/Downloads/MySQL-8.2/mysql-8.2.0-winx64.msi

```
pkg:generic/mysql@8.2.0?download_url=https://dev.mysql.com/get/Downloads/MySQL-8.2/mysql-8.2.0-winx64.msi
```

**Comment by bureado** [comment]
@prabhu re. Winget, interesting. I don't think `purl` is too opinionated on what qualifies as a type, but for the sake of argument: I don't think a Winget manifest can always be reduced 1:1 to a `generic` type, because:

- `Installers` [is an array](https://github.com/microsoft/winget-cli/blob/b96598a6038ac0fb0f58cf935eb466c4484dea3c/schemas/JSON/manifests/v1.0.0/manifest.singleton.1.0.0.json#L525) although I'm not clear if you can have two `Installers` with the same `Scope` and `Architecture`
-  Manifests allow `InstallerSwitches` and other behaviors of potential security interest (such as `Protocols` and `FileExtensions`) which aren't necessarily inherited from the defaults of their `generic` counterpart

I honestly don't know if this is enough to justify a type (it's up to that community to weigh in!) but I also see why a given manifest can't be fully represented with the `generic` for one of their backing installers.

**Comment by prabhu** [comment]
The mysql manifest is [here](https://github.com/microsoft/winget-pkgs/blob/4c926421de9eb26b07368665a69abc1b5d0602e5/manifests/o/Oracle/MySQL/8.0.33/Oracle.MySQL.installer.yaml#L28). To uninstall or update a package installed by winget, the real msi based installer or uninstaller is used. Users can also uninstall from Add/Remove programs. So Winget is only a package manager that is invoking other commands for installation/uninstallation/updates.

To me, this resembles Ubuntu PPA, which supports easy distribution of .deb packages. We use deb as the type for even PPA-distributed packages. To remove these packages, we use apt remove command.

vsix (Visual studio extension) is a package even though it only wraps a bunch of npm packages. Why? Because the vsix is distributed, installed, updated, and removed independently without involving the os package manager.

**Comment by bureado** [comment]
In keeping with the intent of @rhdesmond, let me recap. It sounds like this issue is not about OS identifiers (@prabhu suggested to use the `swid` type) but more about _type_ coverage for Windows systems:

1. Chocolatey is one of those, and it needs to be defined.
2. Generic installers for Windows could likely use `generic`.
3. Store is an open question.
4. And then we had the discussion on Winget.

@prabhu, I think I now agree Winget is not a type. But maybe for other reasons than what you specifically argued. I think it's great that we can persist some perspectives here for whatever proposals come after.

First, I agree that a purl type is not defined by the package manager (tool) This has less to do with the merits of the tool (you can say that `apt-get` is also "only" a package manager that is invoking other commands, like `dpkg-deb` and `install`) and more with the fact that the nature of the package is not mutated by the client, even if the _state_ of the package installation is mutated by the client (in the case of Winget, with switches and locales; in the case of `apt-get`, with `debconf`, maintainer scripts and other local execution hooks and triggers)

I also think that `pkg:deb/ubuntu/dpkg@1.19.0.4?arch=amd64` obtained from `archive.ubuntu.com` is different than obtained from a random PPA. But the `deb` type accommodates for that with `repository_url`.

_Note: I also think if I had a Windows software inventory, I wouldn't just like to see the applications installed, but the installation method - and if it was Winget, to know which manifest it came from. It looks like `winget` can already do this with `winget list` in the `Source` attribute. At any rate, this isn't a problem that `purl` can solve by enforcing an identifier on Winget. Instead, maybe `purl` can help represent an array of `pkg:generic` that make up a Winget manifest._

The fact that Winget calls something else to install the software is not a definitive argument (against it being a type) for me either. Why entertain Helm charts or Stacksmith then? They would all be `purl` identifiers that point to other `purl` identifiers, and not because of composability but because of how distributions work. This semantic transitivity applies to many other types (and in my opinion should absolutely be considered as a factor by implementers, see https://github.com/ossf/wg-securing-critical-projects/issues/74) such as `apk` or `deb`. But I think what differentiates those from Winget is that in Winget, the manifest author does not have room to vendor the installer.

In addition to the ability to vendoring an upstream, purl-representable package, there are likely other considerations for types including the existence of a publishing marketplace, for example. Thanks for the additional perspectives that are helpful for me to expand and complement how I'm thinking about this.. Fascinating discussion, thanks for entertaining it.

**Comment by prabhu** [comment]
Like sex, we all understand purl differently but eventually get there :)

I am happy with using "Is it possible to vendor the installer" as a criterion.

**Comment by chrisrodrigue** [comment]
I fully agree with specifying a type for the [Windows Package Manager (winget)](https://github.com/microsoft/winget-cli). 

It is the official CLI package management tool for Windows and I believe that it ships with Windows 10/11 now. The manifests are maintained in the [winget-pkgs](https://github.com/microsoft/winget-pkgs) repo and it supports MSIX/MSI/APPX/EXE package types.

I think that [winget's `PackageIdentifier`](https://github.com/microsoft/winget-pkgs/blob/6d74648e53b5b7a134a65010c1c82253542d25b5/manifests/m/Microsoft/VCRedist/2012/x64/11.0.61030.0/Microsoft.VCRedist.2012.x64.installer.yaml#L4) YAML key maps most appropriately to the PURL `name` field. It is case-insensitive.

## Example

`pkg:winget/<namespace>/<name>@<version>?<qualifiers>#<subpath>`

`pkg:winget/microsoft.vcredist.2012.x86@11.0.61030.0`

It might be reasonable for one to expect other PURL components to work:

`pkg:winget/microsoft/vcredist.2012@11.0.61030.0?os=windows&arch=x86`

(`os` might be redundant here but I think it would still be smart to have it)

**Comment by matt-phylum** [comment]
If the package manager is <i>win</i>get and the OS is always going to be windows, it seems unnecessary to introduce an os qualifier. If later on winget gets support for other operating systems, os could be defined at that time with a default of windows so the meanings of existing purls would remain unchanged.

The publisher name being split out into the namespace field in the second example is optional. It's mandatory for packages in the winget-pkgs repo to have a publisher, but the winget-pkgs repo is not the only source of packages that can be installed by the winget cli. The [schema used by the winget CLI](https://github.com/microsoft/winget-cli/blob/master/schemas/JSON/manifests/v1.9.0/manifest.version.1.9.0.json) only requires that packages have an ID. Winget comes with a second preinstalled package source where none of the package IDs contain `.`.
